### PR TITLE
[Feature]: MILAN supported

### PR DESCRIPTION
**Pull request by YuanLiuuuuuu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

MILAN 

**linear prob and fine-tuning of the official repo: 78.9 and 85.2**
**linear prob and fine-tuning of mmselfsup: 78.9 and 85.3**

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/benchmarks/classification/imagenet/vit-base-p16_linear-8xb2048-coslr-100e_in1k.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+_base_ = 'vit-base-p16_linear-8xb2048-coslr-90e_in1k.py'
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/imagenet/vit-base-p16_linear-8xb2048-coslr-100e_in1k.py]: this file is duplicated? there is one same config in milan/classfification

File: configs/selfsup/milan/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+_base_ = 'vit-base-p16_ft-8xb128-coslr-100e_in1k.py'
```
Comment (fangyixiao18) [inline configs/selfsup/milan/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan.py]: please check the validation of the base config path

File: configs/selfsup/milan/classification/vit-base-p16_linear-8xb2048-coslr-100e_in1k.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+_base_ = 'vit-base-p16_linear-8xb2048-coslr-90e_in1k.py'
```
Comment (fangyixiao18) [inline configs/selfsup/milan/classification/vit-base-p16_linear-8xb2048-coslr-100e_in1k.py]: also this base file path

File: configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+_base_ = '../mae/mae_vit-base-p16_8xb512-amp-coslr-400e_in1k.py'
```
Comment (fangyixiao18) [inline configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py:1]: better to reduce the inheritance

File: mmselfsup/models/backbones/milan_vit.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    """
+
+    def attention_masking(
+        self, x: torch.Tensor, mask_ratio: float, importance: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor,
+               torch.Tensor, torch.Tensor]:
+        """Generate attention mask for MILAN.
+
+        This is what is different from MAEViT, which uses random masking.
+        attention masking generates attention mask for MILAN, according to
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/milan_vit.py]:
```suggestion
        Attention masking generates attention mask for MILAN, according to
```

File: configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            dict(type='Constant', layer='LayerNorm', val=1.0, bias=0.0)
+        ]),
+    head=dict(
+        _delete_=True,
+        type='MILANPretrainHead',
+        loss=dict(type='MILANReconstructionLoss')),
+    target_generator=dict(
+        type='CLIPGenerator',
+        tokenizer_path=  # noqa
+        '/mnt/cache/liuyuan/research/MILAN/work_dirs/clip_vit_base_16.pth.tar'
```
Comment (fangyixiao18) [inline configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py]: change to https link

File: mmselfsup/models/target_generators/clip_generator.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    This module is used to generate target features in masked image modeling.
+
+    Args:
+        tokenizer_path (str): The path of the checkpoint of CLIP.
+    """
+
+    def __init__(self, tokenizer_path: str) -> None:
+        super().__init__()
+        self.tokenizer = build_clip_model(torch.load(tokenizer_path), False)
```
Comment (fangyixiao18) [inline mmselfsup/models/target_generators/clip_generator.py]: better to support to load from url
using mmengine.runner.checkpoint.load_checkpoint is better.

File: mmselfsup/models/utils/clip.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (fangyixiao18) [inline mmselfsup/models/utils/clip.py:1]: add url or copyright of clip

File: mmselfsup/models/utils/clip.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        finetune,
+        average_targets,
+    )
+
+    for key in ['input_resolution', 'context_length', 'vocab_size']:
+        if key in state_dict:
+            del state_dict[key]
+
+    msg = model.load_state_dict(state_dict, strict=False)
+    print('Load CLIP model:', msg)
```
Comment (fangyixiao18) [inline mmselfsup/models/utils/clip.py]: use logger instead of print

**Comment by YuanLiuuuuuu** [comment]
image

File: mmselfsup/models/utils/transformer_blocks.py [hunk 10]
```diff
@@ -13,6 +13,175 @@
...
+                encoder with shape N x L_v x C.
+            ids_restore (torch.Tensor): The ids of all tokens in the original
+                image with shape N x L.
+
+        Returns:
+            torch Tensor: Output features with shape N x L x C.
+        """
+        x_ = torch.cat([visible_tokens[:, 1:, :], x], dim=1)
+        assert x_.shape[1] == ids_restore.shape[1]
+        x_ = torch.gather(
```
Comment (fangyixiao18) [inline mmselfsup/models/utils/transformer_blocks.py:89]: indicate the gather op in docstring, which is also different from _MultiheadAttention.

File: configs/selfsup/milan/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+and a semantic aware mask sampling mechanism, which further advance the
+transfer performance of the pretrained model. Experimental results demonstrate
+that MILAN delivers higher accuracy than the previous works. When the masked
+autoencoder is pretrained and finetuned on ImageNet-1K dataset with an input
+resolution of 224×224, MILAN achieves a top-1 accuracy of 85.4% on ViTB/16, surpassing previous state-of-the-arts by 1%. In the downstream semantic
+segmentation task, MILAN achieves 52.7 mIoU using ViT-B/16 backbone on
+ADE20K dataset, outperforming previous masked pretraining results by 4 points.
+
+<div align="center">
+<img src="https://user-images.githubusercontent.com/30762564/205210369-41a65c4c-bcd4-4147-91ea-c6c9061ab455.png" width="40%"/>
```
Comment (fangyixiao18) [inline configs/selfsup/milan/README.md]: The image is kind of small and not very clear in readme

File: configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py (hunk 4 above)
Comment (YuanLiuuuuuu) [reply configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py:1]: The inheritance level is the same as those configs in MAE.

File: mmselfsup/models/utils/transformer_blocks.py (hunk 10 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/utils/transformer_blocks.py:89]: The main distinction is this module will not update the visible tokens, while this ``gather`` op is used to implement this func.

File: mmselfsup/models/utils/clip.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+class QuickGELU(nn.Module):
+    """A faster version of GELU."""
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward function."""
+        return x * torch.sigmoid(1.702 * x)
+
+
+class ResidualAttentionBlock(nn.Module):
+    """Residual Attention Block (RAB).
```
Comment (fangyixiao18) [inline mmselfsup/models/utils/clip.py:31]: Better to indicate the difference with the general transformer block

File: mmselfsup/models/utils/clip.py [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+            x = x + x_
+            x = x + self.mlp(self.ln_2(x))
+            return x, attention
+        else:
+            x = x + self.attention(self.ln_1(x))
+            x = x + self.mlp(self.ln_2(x))
+            return x
+
+
+class Transformer(nn.Module):
```
Comment (fangyixiao18) [inline mmselfsup/models/utils/clip.py:97]: The name here is kind of confusing, maybe add more docstring

File: configs/selfsup/milan/README.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+      <td><a href='https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/configs/selfsup/milan/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan.py'>config</a> |<a href='https://download.openmmlab.com/mmselfsup/1.x/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k/vit-base-p16_ft-8xb128-coslr-100e_in1k/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan_20221129-74ac94fa.pth'> model </a>| <a href='https://download.openmmlab.com/mmselfsup/1.x/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k/vit-base-p16_ft-8xb128-coslr-100e_in1k/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan_20221125_031826.json'>log</a></td>
+	</tr>
+</tbody>
+</table>
+## Citation
```
Comment (tonysy) [inline configs/selfsup/milan/README.md:73]:
```suggestion

## Citation
```

File: configs/selfsup/milan/README.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+segmentation task, MILAN achieves 52.7 mIoU using ViT-B/16 backbone on
+ADE20K dataset, outperforming previous masked pretraining results by 4 points.
+
+<div align="center">
+<img src="https://user-images.githubusercontent.com/30762564/205210369-41a65c4c-bcd4-4147-91ea-c6c9061ab455.png" width="80%"/>
+</div>
+
+## Prerequisite
+
+Create a new folder `milan_ckpt` under the root directory and download the
```
Comment (tonysy) [inline configs/selfsup/milan/README.md]: root directory is confusing, doest mean  in home folder of mmselfsup?

File: docs/en/model_zoo.md [hunk 16]
```diff
@@ -393,5 +393,16 @@ ImageNet has multiple versions, but the most commonly used one is ILSVRC 2012. T
       <td>/</td>
       <td><a href='https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/configs/benchmarks/classification/imagenet/vit-base-p16_ft-8xb256-coslr-100e_in1k.py'>config</a> | <a href='https://download.openmmlab.com/mmselfsup/1.x/maskfeat/maskfeat_vit-base-p16_8xb256-amp-coslr-300e_in1k/vit-base-p16_ft-8xb256-coslr-100e_in1k/vit-base-p16_ft-8xb256-coslr-100e_in1k_20221028-5134431c.pth'>model</a> | <a href='https://download.openmmlab.com/mmselfsup/1.x/maskfeat/maskfeat_vit-base-p16_8xb256-amp-coslr-300e_in1k/vit-base-p16_ft-8xb256-coslr-100e_in1k/vit-base-p16_ft-8xb256-coslr-100e_in1k_20221026_105344.json'>log</a></td>
 	</tr>
+  <tr>
```
Comment (tonysy) [inline docs/en/model_zoo.md]: - I recommand to follow the MMClassification's style to present the checkpoints. https://mmclassification.readthedocs.io/en/dev-1.x/modelzoo_statistics.html
- We can create an issue to backlog @fangyixiao18

File: mmselfsup/models/losses/milan_loss.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+import torch
+from mmengine.model import BaseModule
+
+from mmselfsup.registry import MODELS
+
+
+@MODELS.register_module()
+class MILANReconstructionLoss(BaseModule):
```
Comment (tonysy) [inline mmselfsup/models/losses/milan_loss.py]: Refactor it by using cosine similarity loss

**Review by tonysy — CHANGES_REQUESTED** [review]
As inline comments

File: mmselfsup/models/utils/clip.py (hunk 13 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/utils/clip.py:97]: I add docstring for this module, but keeping the name unchanged is better, to keep consistent with that in CLIP.

File: configs/selfsup/milan/README.md (hunk 15 above)
Comment (YuanLiuuuuuu) [reply configs/selfsup/milan/README.md]: Currently, we support loading a ckpt from a url, so this is no need to download this ckpt manually. I will remove this part.

File: mmselfsup/models/losses/milan_loss.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (fangyixiao18) [inline mmselfsup/models/losses/milan_loss.py]: This file can be removed

File: tests/test_models/test_algorithms/test_milan.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+loss = dict(type='MILANReconstructionLoss')
+head = dict(type='MILANPretrainHead', loss=loss)
+# target_generator = dict(
+#     type='CLIPGenerator',
+#     tokenizer_path=  # noqa
+#     'milan_ckpt/clip_vit_base_16.pth.tar')
+
+
+@pytest.mark.skipif(platform.system() == 'Windows', reason='Windows mem limit')
+def test_mae():
```
Comment (fangyixiao18) [inline tests/test_models/test_algorithms/test_milan.py]: test_milan

File: configs/selfsup/milan/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan.py [hunk 20]
```diff
@@ -0,0 +1,12 @@
+_base_ = '../../../benchmarks/classification/imagenet/vit-base-p16_ft-8xb128-coslr-100e_in1k.py'  # noqa: E501
```
Comment (tonysy) [inline configs/selfsup/milan/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k-milan.py]: Update filename without milan

**Review by tonysy — CHANGES_REQUESTED** [review]
Update filename and this PR is LGTM

**Review by fangyixiao18 — COMMENTED** [review]
LGTM

**Review by tonysy — APPROVED** [review]
LGTM
